### PR TITLE
Enable memory-mapped stress log on cross platform

### PR DESCRIPTION
**Pull request by cshung**
This change enables the memory-mapped stress log to be used on Linux platforms.

**Changes:**
- **PAL Changes**
-   - Implemented `InterlockedAdd` and `InterlockedAdd64`.
  - Introduces `PAL_CopyModuleData` and implemented it for Linux and OSX.
  - Extend `MapGrowLocalFile` to support file size > 2**32.
- **Runtime Changes:**
  - Introduce the `StressLog::s_memoryMapped` field to denote whether the current process is running the memory-mapped stress log file mode. Previously, we were using `StressLog::s_LogChunkHeap == NULL` to denote that case, but that's a Windows-only thing.
  - Changed various `wchar_t` to `WCHAR`
  - Use `PAL_CopyModuleData` to copy the module data in the cross-plat case.
- **Tools Changes:**
  - Changed various `wchar_t` to `char`, as the strings are used mostly for format string and they're ASCII. 
  - Initialize the PAL for cross-platform
  - Eliminated the dynamic linking of the plugin and statically link instead.
  - There is no support for a guaranteed base address for mapping files to memory on cross-platform, so I implemented an address translation scheme so that we can decode even when the file is mapped to an alternative base address.

There are many small changes everywhere to get all the compilers happy, mostly type casting and format string changes.

**Highlights:**
- The runtime part can be compiled with no error on all platforms.
- The analyzer tool can read the generated StressLog file on (Windows x64, Linux x64, OSX x64) and perform format string filtering.

**Lowlights:**
- I dropped the support for dynamically loading plugins while the analyzer is running, getting dynamic loading to work across platforms is more work than the benefits with this reloading.
- On Linux, we have got a crash if the stress log file is already there.
- I dropped the support for Alpine Linux, the compiler is mysteriously crashing.

**TODO**
- Investigate and fix the Linux crash - but it is easy to workaround.

### CODE REVIEW

File: src/coreclr/utilcode/stresslog.cpp [hunk 1]
```diff
@@ -185,13 +188,19 @@ static LPVOID CreateMemoryMappedFile(LPWSTR logFilename, size_t maxBytesTotal)
     }
 
     size_t fileSize = maxBytesTotal;
+    // TODO: PAL does not support mapping larger than 32 bit
```
Comment (janvorli) [inline src/coreclr/utilcode/stresslog.cpp:191]: It seems that we can fix it in PAL easily. The `ftruncate` Unix function that ends up getting called has `off_t` type of the argument, which is 64 bit even for our 32 bit builds now. So it seems that modifying the type of the `NewSize`  argument of the `MAPGrowLocalFile` to be `off_t` and updating the `CorUnix::InternalCreateFileMapping` to combine the low / high parts into an `off_t` value and using it in that method instead of just `dwMaximumSizeLow` would do the trick.

Comment (janvorli) [reply src/coreclr/utilcode/stresslog.cpp:191]: And remove the `if (0 != dwMaximumSizeHigh)` where we bail out too.

File: src/coreclr/utilcode/stresslog.cpp [hunk 2]
```diff
@@ -185,13 +188,19 @@ static LPVOID CreateMemoryMappedFile(LPWSTR logFilename, size_t maxBytesTotal)
...
     if (hMap == NULL)
     {
         return nullptr;
     }
 
+#ifdef HOST_WINDOWS
     return MapViewOfFileEx(hMap, FILE_MAP_ALL_ACCESS, 0, 0, fileSize, (void*)0x400000000000);
+#else
+    // TODO: PAL does not support mapping with a specified base address
+    return MapViewOfFileEx(hMap, FILE_MAP_ALL_ACCESS, 0, 0, fileSize, nullptr);
```
Comment (janvorli) [inline src/coreclr/utilcode/stresslog.cpp:202]: While we possibly might be able to add support for this, it would not make sense to pick any static address on Unix. Moreover, it would be just a hint anyways. So let's do it the way you did, you can remove the TODO from this.
A nit - it would be nicer to make a `#define` for the base address that would be platform specific and leave the code the same.

Comment (cshung) [reply src/coreclr/utilcode/stresslog.cpp]: With some experimentation, I believe `StressLogAnalyzer` works by mapping the file back to exactly where it was. The reason is that once it map the file back to memory, it started tracing pointers, and these pointers could be pointing into the mapped area. For example, we are dereferencing `log.t` here:

https://github.com/dotnet/runtime/blob/8fbf206d0e518b45ca855832e8bfb391afa85972/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin/StressLogPlugin.cpp#L1145

I wonder if we could make it work by sending the base address parameter to the first parameter of `mmap`, they seem to serve the same purpose.

I understand there is a possibility for the API to fail, it looks like I have to handle the relocation of those pointers regardless :(

Comment (janvorli) [reply src/coreclr/utilcode/stresslog.cpp]: The first parameter to mmap is just a hint. The mmap will completely ignore it when some SELinux security measures are on and it would also pick a different address if there already is a mapping overlapping the range. I guess that in Windows address space, we can rely on this address range being always (?) available, but we don't have that luxury on Unix. So it would be better to not to rely on a specific base address.

File: src/coreclr/inc/stresslog.h [hunk 3]
```diff
@@ -772,7 +797,7 @@ class ThreadStressLog {
 
     BOOL IsValid () const
     {
-        return chunkListHead != NULL && (!curWriteChunk || curWriteChunk->IsValid ());
+        return chunkListHead != NULL && (!curWriteChunk || StressLog::TranslateMemoryMappedPointer(curWriteChunk)->IsValid ());
```
Comment (cshung) [inline src/coreclr/inc/stresslog.h:800]: This needs a bit of explanation. The address translation is needed only for the stress log analyzer and this seems to be the only one that the stress log analyzer calls into and performs a pointer dereferencing.

**Comment by Maoni0** [comment]
do you observe any perf difference between mapping to a fixed address and having to do address translation on windows? presumably it was a fixed address for perf reasons.

**Comment by cshung** [comment]
> do you observe any perf difference between mapping to a fixed address and having to do address translation on windows? presumably it was a fixed address for perf reasons.

I do not observe a difference in performance when I am testing, but I wasn't testing on particularly large files either. 

I am not sure that the fixed base address was for perf. It is just a lot easier to do if we don't have to translate addresses. Unfortunately, we have to for cross-platform.

The address translation only happens during `StressLogAnalyzer` when we read the file. There is no runtime performance impact.

The majority of the address translation is to handle chunks. For every 16k data we read from the file, we need to do a few address translations to get to the next one. Most of them are debug only for assertions. So I don't foresee a performance problem there.

We have been doing another address translation for getting to the format string for each stress log message already.

**Comment by cshung** [comment]
@PeterSolMS, can you please review?
@janvorli, can you please review again? The new version has support for OSX.

File: src/coreclr/pal/inc/pal.h [hunk 4]
```diff
@@ -3252,6 +3257,37 @@ FORCEINLINE void PAL_ArmInterlockedOperationBarrier()
 #endif // HOST_ARM64
 }
 
+/*++
+TODO: andrewau, comments
```
Comment (janvorli) [inline src/coreclr/pal/inc/pal.h:3261]: Can you please add the comments as indicated by your TODO?

File: src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp [hunk 5]
```diff
@@ -80,14 +95,15 @@ const char *getFacilityName(DWORD_PTR lf)
         #include "../../../inc/loglf.h"
         { LF_ALWAYS, "LF_ALWAYS" }
     };
-    static char buff[1024] = "`";
+    const int buff_length = 1024;
```
Comment (janvorli) [inline src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp:98]: A nit - I wonder if you have used the `buff_length` for something before and then changed stuff. It seems unnecessary when it is used just to create the array. Similar for the `format_length` and `name_buffer_length` below.

File: src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp [hunk 6]
```diff
@@ -162,8 +179,8 @@ void formatOutput(struct IDebugDataSpaces* memCallBack, ___in FILE* file, __inou
                         // Print the string up to that point
                     c = *ptr;
                     *ptr = 0;       // Terminate the string temporarily
-                    vfprintf(file, format, (va_list)args);
-                    *ptr = c;       // Put it back
+                    fprintf(file, format, args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], args[8], args[9], args[10]);
```
Comment (janvorli) [inline src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp:182]: Why was this change needed? PAL has `vfprintf` implementation.

File: src/coreclr/utilcode/stresslog.cpp [hunk 7]
```diff
@@ -162,12 +165,12 @@ static LPVOID CreateMemoryMappedFile(LPWSTR logFilename, size_t maxBytesTotal)
 
         // append the string representation of the PID
         DWORD pid = GetCurrentProcessId();
-        wchar_t pidStr[20];
-        _itow(pid, pidStr, 10);
-        wcscat_s(fileName, pidStr);
+        WCHAR pidStr[20];
+        _itow_s(pid, pidStr, 20, 10);
```
Comment (janvorli) [inline src/coreclr/utilcode/stresslog.cpp:169]: Can you use something like ARRAY_SIZE here for the 20 to eliminate the possibility of future mismatched `pidStr`? We use various names for that in the runtime, we have _countof or COUNTOF and both of them used at various places in utilcode.

File: src/coreclr/tools/CMakeLists.txt [hunk 8]
```diff
@@ -4,3 +4,6 @@ if (CLR_CMAKE_TARGET_WIN32 AND NOT CLR_CMAKE_CROSS_ARCH)
 
   install(EXPORT dactabletools DESTINATION dactabletools COMPONENT crosscomponents)
 endif()
+if ((NOT CLR_CMAKE_TARGET_ALPINE_LINUX) AND (CLR_CMAKE_TARGET_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_ARM64))
+  add_subdirectory(StressLogAnalyzer)
```
[the remark this answers is not in the document]
Comment (cshung) [reply src/coreclr/tools/CMakeLists.txt:8]: The reason why Alpine is disabled is that it doesn't work there.

I did a couple of experiments with it:
- On CI - the build log gives a rather meaningless error on the build log, saying something is killed - not even sure what was killed
- Locally, I tried to build using Docker on a remote physical Linux machine. During the build, the machine is disconnected from SSH and no longer connects. The hypothesis is that the machine crashed, and eventually that needs a physical reboot to recover.

We don't have an immediate need for getting the memory-mapped stress log on Alpine right now, so we decided to scope out Alpine for now. As need arises, we can then go investigate what is going on there.

File: src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp (hunk 6 above)
Comment (cshung) [reply src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp:182]: The `vfprintf` function works. The problem is that on Linux, the cast from `void**` to `va_args` is forbidden.

Comment (janvorli) [reply src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp:182]: Ah, I haven't noticed that we are passing in args as void**. The failure to cast on Unix makes sense, varargs is not a pointer, but rather a data structure there.

File: src/coreclr/tools/StressLogAnalyzer/StressLogAnalyzer.cpp [hunk 9]
```diff
@@ -51,16 +69,30 @@ int ParseCommandLine(wchar_t* s, wchar_t** argv, int maxArgc)
...
 }
 
-int wmain(int argc, wchar_t *argv[])
+int ProcessStressLog(void* baseAddress, int argc, char* argv[]);
+
+int main(int argc, char *argv[])
 {
-    if (argc < 2 || wcscmp(argv[1], L"-?") == 0)
+#ifdef HOST_UNIX
+    int exitCode = PAL_InitializeDLL();
```
Comment (janvorli) [inline src/coreclr/tools/StressLogAnalyzer/StressLogAnalyzer.cpp:77]: This should call `PAL_Initialize` instead, the `PAL_InitializeDLL` is meant for shared libraries.

**Review by janvorli — APPROVED** [review]
LGTM modulo the nit.

File: src/coreclr/tools/StressLogAnalyzer/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,14 @@
+set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreaded)
+add_executable(StressLogAnalyzer StressLogAnalyzer.cpp StressLogDump.cpp StressLogPlugin.cpp)
```
Comment (am11) [inline src/coreclr/tools/StressLogAnalyzer/CMakeLists.txt:2]:
```suggestion
add_executable_clr(StressLogAnalyzer StressLogAnalyzer.cpp StressLogDump.cpp StressLogPlugin.cpp)
```

File: src/coreclr/tools/StressLogAnalyzer/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    target_link_libraries(StressLogAnalyzer
+        ${STATIC_MT_CRT_LIB}
+        ${STATIC_MT_VCRT_LIB}
+    )
+else()
+    target_link_libraries(StressLogAnalyzer
+        coreclrpal
+    )
+endif(CLR_CMAKE_TARGET_WIN32)
+
```
Comment (am11) [inline src/coreclr/tools/StressLogAnalyzer/CMakeLists.txt:14]:
```suggestion
endif(CLR_CMAKE_TARGET_WIN32)

install_clr(TARGETS StressLogAnalyzer DESTINATIONS . COMPONENT runtime)
```

**Review by am11 — COMMENTED** [review]
For the TODO
> The built binary should be placed under the bin directory, right now it is under obj.

**Comment by cshung** [comment]
> For the TODO
> 
> > The built binary should be placed under the bin directory, right now it is under obj.

Thank you!

**Comment by Maoni0** [comment]
I didn't mean this change was done for perf (and obviously it does not affect runtime perf when it's not on) but I'm still curious about perf differences (tooling perf matters too). the reason why I was asking was because we used stress log for dprintf for perf reasons so it would be good to not degraded it much. 

> We have been doing another address translation for getting to the format string for each stress log message already.

ahh, then I wouldn't expect you'd see much perf diff at all with your change. that's good to know.
